services/user/Chainmail/service/src/lib.rs

```diff
@@ -158,29 +47,100 @@ mod service {
         EventsSvc::call().addIndex(DbId::HistoryEvent, SERVICE, MethodNumber::from("sent"), 1);
     }
 
+    /// Send message
+    /// by emiting a `sent` event
     #[action]
     fn send(receiver: AccountNumber, subject: String, body: String) {
         check(
             AccountsSvc::call().exists(receiver),
             &format!("receiver account {} doesn't exist", receiver),
         );
+        let datetime = TransactSvc::call().currentBlock().time;
         Wrapper::emit()
             .history()
-            .sent(get_sender(), receiver, subject, body);
+            .sent(get_sender(), receiver, subject, body, datetime);
     }
 
+    /// Archive message
+    /// by emiting an `archived` event
     #[action]
     fn archive(msg_id: u64) {
+        let saved_messages_table = SavedMessagesTable::new();
+        if let Some(rec) = saved_messages_table.get_index_pk().get(&msg_id) {
+            unsave(
+                msg_id,
+                rec.sender,
+                rec.subject,
+                rec.body,
+                rec.datetime.seconds,
+            );
+        }
+
         Wrapper::emit()
             .history()
             .archive(get_sender().to_string() + &msg_id.to_string());
     }
 
+    /// Save a message
+    /// Events can be pruned by nodes. Since Chainmail relies on events to "store" messages,
+    /// `save` copies a message into state where it can be stored indefinitely at the user's expense.
+    #[action]
+    fn save(
+        msg_id: u64,
+        receiver: AccountNumber,
+        sender: AccountNumber,
+        subject: String,
+        body: String,
+        datetime: u32,
+    ) {
+        check(
+            get_sender() == receiver,
```

Comment:
@James-Mart If I look up the message being referenced (by msg_id) and confirm that message has the other details specified int he args to the fn, then this check work, right? because we always know get_sender() is valid, and if we know the msg being referenced is valid, then this check works, right? So I could add the check for msg-by-id. If I can't find the msg, we should fail because either 1) this is a fake message being "saved" or 2) the message was pruned and so can't be saved.

Reply:
Events are in a write-only table.

Anyway, nothing wrong with people saving fake messages, since they are paying. And (as mentioned) we can add the ability to verify saved messages later.